winpr/libwinpr/environment/test/TestEnvironmentGetEnvironmentStrings.c

```diff
@@ -26,7 +24,7 @@ int TestEnvironmentGetEnvironmentStrings(int argc, char* argv[])
 			_tprintf(_T("test failed: return %d\n"), rc);
 			goto fail;
 		}
-		length = _tcslen(p);
+		const size_t length = _tcsnlen(p, (size_t)rc);
```

Comment:
The size specifier for this `_tcsnlen` call is wrong. `p` is in no relation to `rc` (its size is only part of it). Since we do not know or can estimate its size (and `GetEnvironmentStrings` is defined to be correctly terminated) I'd stick to `_tcslen` here.

Reply:
@mfleisz `rc` is always `tcslen(p) + 2` (newline + '\0') so the limit does make sense.
this way you get early termination of `tcsnlen` if the terminators are missing - well, then it most likely already failed with `_sntprintf`, but none the less)

Reply:
strnlen is not a replacement for strlen

Reply:
This is actually wrong since in this case `rc` will be always greater than the length of `p` so we do not gain anything from using `strnlen`.

I think [this](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/strnlen-strnlen-s?view=msvc-170#remarks) gives a pretty good explanation of when to use strnlen - and it begins with "**strnlen** is not a replacement for **strlen**". Applied to our cause I think `strnlen` is misused as we are not limiting the strings length but actually extending it with `rc`.

Reply:
Additional note: `snprintf` does not count the trailing `\0`

Reply:
@mfleisz true. did mix up that part, but the rest is still valid (`rc > tcslen(p)` is still true)

Reply:
@mfleisz ok, found a way.
https://devblogs.microsoft.com/oldnewthing/20100203-00/?p=15083 limits the whole block to `32768` char.
so, a sensible upper bound for a single variable (although maybe too big) is `32768` char

Reply:
@mfleisz as for your link, that is exactly the reason I try to eliminate these.
While some instances here might also work with `strlen` et al, most of `FreeRDP` parses data from (untrusted) network sources.
So, eliminating these altogether from our codebase (and we are not talking about a lot of instances here) gives us the freedom to tighten the pr checker further (e.g. fail if such functions are introduced again)